### PR TITLE
Fix issue related to generalizing isolated edge

### LINKED ISSUE #2744: Preserver order breaks pipeline if hashJoin is used
If pipeline includes `hashJoin` stage (or `hashJoin2`, `innerHashJoin`, `innerHashJoin2`) and `preservedOrder` is set to `true` and partitioned source is used (tested with `mapJournal`) then job fails with:
```
Exception in ProcessorTasklet{mapJournalSource(sourceMap)#1}
java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/java.util.Arrays$ArrayList.get(Arrays.java:4164)
	at com.hazelcast.jet.impl.util.CircularListCursor.value(CircularListCursor.java:57)
	at com.hazelcast.jet.impl.execution.OutboundCollector$RoundRobin.offer(OutboundCollector.java:121)
	at com.hazelcast.jet.impl.execution.OutboxImpl.doOffer(OutboxImpl.java:251)
	at com.hazelcast.jet.impl.execution.OutboxImpl.offerInternal(OutboxImpl.java:139)
	at com.hazelcast.jet.impl.execution.OutboxImpl.offer(OutboxImpl.java:94)
	at com.hazelcast.jet.core.AbstractProcessor.tryEmit(AbstractProcessor.java:309)
	at com.hazelcast.jet.core.AbstractProcessor.emitFromTraverser(AbstractProcessor.java:408)
	at com.hazelcast.jet.core.AbstractProcessor.emitFromTraverser(AbstractProcessor.java:421)
	at com.hazelcast.jet.impl.connector.StreamEventJournalP.emitResultSet(StreamEventJournalP.java:207)
	at com.hazelcast.jet.impl.connector.StreamEventJournalP.complete(StreamEventJournalP.java:192)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.complete(ProcessorTasklet.java:467)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:367)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:242)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366)
	at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:807)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

Steps to reproduce:
```
Pipeline p = Pipeline.create();
StreamStage<Map.Entry<Integer, Integer>> source
        = p.readFrom(Sources.<Integer, Integer>mapJournal("sourceMap", JournalInitialPosition.START_FROM_OLDEST))
                .withoutTimestamps();
BatchStage<Map.Entry<Integer, String>> enr = p.readFrom(Sources.map("enrichMap"));
StreamStage<Map.Entry<Integer, String>> hashJoin = source.hashJoin(
        enr,
        joinMapEntries(t -> t.getKey()),
        (t, u) -> {
            return entry(t.getKey(), t.getValue() + u);
        }
);
hashJoin.writeTo(Sinks.logger());
p.setPreserveOrder(true);
JetConfig config = new JetConfig();
EventJournalConfig eventJournalConfig = config.getHazelcastConfig().getMapConfig("sourceMap").getEventJournalConfig();
eventJournalConfig.setEnabled(true);
JetInstance jet = Jet.newJetInstance(config);
try {
    IMap<Integer, String> eMap = jet.getMap("enrichMap");
    eMap.put(0, "a");
    eMap.put(1, "b");
    Job job = jet.newJob(p);
    Thread.sleep(5000);
    IMap<Integer, Integer> map = jet.getMap("sourceMap");
    for (int i = 0; i < 100; i++) {
        map.put(0, i);
        map.put(1, i);
    }
    Thread.sleep(5000);
    job.cancel();
    Thread.sleep(1000);
    System.out.println("DONE");
} finally {
    jet.shutdown();
}
```

### PR DESCRIPTION
**Pull request by ufukyilmaz**
We extended the definition of the isolated edge within the scope of event reordering PR:#2522 , but there was error in its implementation and it was breaking the pipelines where hashJoin*, innerJoin* etc. is used. This PR solves this issue. 

Fixes #2744

Checklist:
- [x] Labels and Milestone set

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3796/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3796/jet-oss-pr-builder-3796.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/MockLoggingFactory.java:[67,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:testCompile (default-testCompile) on project hazelcast-jet-core: Compilation failure
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/MockLoggingFactory.java:[67,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/MockLoggingFactory.java:[67,5] method does not override or implement a method from a supertype
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3797/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3797/jet-oss-pr-builder-3797.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/MockLoggingFactory.java:[67,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:testCompile (default-testCompile) on project hazelcast-jet-core: Compilation failure
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/MockLoggingFactory.java:[67,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/MockLoggingFactory.java:[67,5] method does not override or implement a method from a supertype
--------------------------
</details>
</pre>

**Comment by ufukyilmaz** [comment]
verify

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3801/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3801/jet-oss-pr-builder-3801.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   StreamKafkaPTest.when_partitionAddedWhileJobDown_then_consumedFromBeginning:404->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248->lambda$when_partitionAddedWhileJobDown_then_consumedFromBeginning$8:405 expected:<1> but was:<2>
[INFO] 
[ERROR] Tests run: 35, Failures: 1, Errors: 0, Skipped: 0
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by ufukyilmaz** [comment]
verify

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java [hunk 1]
```diff
@@ -472,13 +474,35 @@ private OutboundEdgeStream createOutboundEdgeStream(EdgeDef edge,
...
+                            return createConveyorArray(downstreamParallelism, queueCount, queueSize);
+                        } else if (upstreamParallelism - remainder == 0) { // downstreamParallelism > upstreamParallelism
+                            return createConveyorArray(downstreamParallelism, queueCount + 1, queueSize);
+                        } else {
+                            return Stream.concat(
+                                    Arrays.stream(createConveyorArray(remainder, queueCount + 1, queueSize)),
+                                    Arrays.stream(createConveyorArray(
+                                            downstreamParallelism - remainder, queueCount, queueSize
+                                    ))).toArray((IntFunction<ConcurrentConveyor<Object>[]>) ConcurrentConveyor[]::new);
+                        }
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java:491]:
```suggestion
                        return Stream.concat(
                                Arrays.stream(createConveyorArray(remainder, queueCount + 1, queueSize)),
                                Arrays.stream(createConveyorArray(
                                        downstreamParallelism - remainder, Math.max(1, queueCount), queueSize
                                ))).toArray((IntFunction<ConcurrentConveyor<Object>[]>) ConcurrentConveyor[]::new);
```

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java [hunk 2]
```diff
@@ -472,13 +474,35 @@ private OutboundEdgeStream createOutboundEdgeStream(EdgeDef edge,
...
+                        .filter(i -> i % upstreamParallelism == processorIndex)
+                        .mapToObj(i -> new ConveyorCollector(localConveyors[i], 0, null))
+                        .toArray(OutboundCollector[]::new);
+            } else {
+                return IntStream.range(0, downstreamParallelism)
+                        .filter(i -> processorIndex % downstreamParallelism == i)
+                        .mapToObj(i -> new ConveyorCollector(localConveyors[i],
+                                processorIndex / downstreamParallelism, null))
+                        .toArray(OutboundCollector[]::new);
+            }
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java:505]:
```suggestion
                return IntStream.range(0, downstreamParallelism)
                        .filter(i -> i % upstreamParallelism == processorIndex % downstreamParallelism)
                        .mapToObj(i -> new ConveyorCollector(localConveyors[i], processorIndex / downstreamParallelism, null))
                        .toArray(OutboundCollector[]::new);
```

**Review by viliam-durina — APPROVED** [review]
Original code was correct, I generalized the formulas so that we don't need to specially handle cases when upstream parallelism is larger/smaller/equal than downstream's.

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3819/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3819/jet-oss-pr-builder-3819.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:checkstyle (default) on project hazelcast-jet-core: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder_2/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java:490: Line is longer than 123 characters (found 126). [LineLength]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3820/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3820/jet-oss-pr-builder-3820.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:checkstyle (default) on project hazelcast-jet-core: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-oss-pr-builder/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java:490: Line is longer than 123 characters (found 126). [LineLength]
--------------------------
</details>
</pre>

**Pull request merged**